### PR TITLE
DPL Analysis: Temporary solution for filtering on multiple variables at once

### PR DESCRIPTION
**Pull request by saganatt**
This would enable to apply a filter to a combination of elements e.g. `tracks1.pt() + tracks2.pt() > cut`.

### CODE REVIEW

File: Framework/Core/include/Framework/ASoA.h [hunk 1]
```diff
@@ -1289,12 +1289,6 @@ class Filtered : public T
 template <typename T>
 using is_soa_filtered_t = typename framework::is_specialization<T, soa::Filtered>;
 
-template <typename T>
-auto filter(T&& t, framework::expressions::Filter const& expr)
```
Comment (saganatt) [inline Framework/Core/include/Framework/ASoA.h]: Seems to be a deprecated remainder, not used anywhere

**Comment by ktf** [comment]
Is this ready to be tested?

**Comment by saganatt** [comment]
This is not yet a working code.
I have talked with Anton about adapting the filters, so I have a better idea how to improve it, although I am not sure if I will manage to make it working the way we want. Full functionality will be achievable more easily once Anton extends the filters with IterationNodes.

**Comment by ktf** [comment]
is this still open?

File: Framework/Core/test/benchmark_ASoAHelpers.cxx [hunk 2]
```diff
@@ -457,4 +457,430 @@ static void BM_ASoAHelpersCombGenTracksFivesMultipleChunks(benchmark::State& sta
 
 BENCHMARK(BM_ASoAHelpersCombGenTracksFivesMultipleChunks)->RangeMultiplier(2)->Range(8, 8 << (maxFivesRange - 1));
 
+//static void BM_ASoAHelpersCombGenSimplePairsFilters(benchmark::State& state)
+//{
+//  // Seed with a real random value, if available
+//  std::default_random_engine e1(1234567891);
+//  std::uniform_real_distribution<float> uniform_dist(0, 1);
+//
```
Comment (ktf) [inline Framework/Core/test/benchmark_ASoAHelpers.cxx:465]: Please do not add commented code.

**Comment by saganatt** [comment]
This is a working but still a very slow code. I am checking now possibilities with SelectionVector in ASoA.h, the combinations themselves make a little overhead, if any.

**Comment by saganatt** [comment]
Rebased on clean PR #3438

File: Framework/Core/test/test_ASoAHelpers.cxx [hunk 3]
```diff
@@ -842,10 +968,192 @@ BOOST_AUTO_TEST_CASE(BlockCombinations)
...
+  for (auto& [t0, t1, t2] : selfCombinations<CombinationsFilteredStrictlyUpperIndexPolicy>(pairsFilter, testsA, testsA, testsA)) {
+    BOOST_CHECK_EQUAL(t0.x(), std::get<0>(expectedStrictlyUpperTriples2[count]));
+    BOOST_CHECK_EQUAL(t1.x(), std::get<1>(expectedStrictlyUpperTriples2[count]));
+    BOOST_CHECK_EQUAL(t2.x(), std::get<2>(expectedStrictlyUpperTriples2[count]));
+    count++;
+  }
+  BOOST_CHECK_EQUAL(count, 27);
+
+  // Order in tuple: (t0 > 5) && (t2 + t1 > 10) && (t4 + t3 > 9)
+  expressions::Filter fivesFilter = (test::x > 5) && (test::x > (test::x * (-1) + 9)) && (test::x > (test::x * (-1) + 4));
```
Comment (saganatt) [inline Framework/Core/test/test_ASoAHelpers.cxx:1146]: Filter issues are most visible here:
- not possible to write e.g. `test::x + test::x > 9`, what makes filter hard to understand
- not intuitive order of iterators in a tuple due to tree structure of the filter. For example if `(test::x > 5)` would be the last condition, it still would be applied to `t0` (same order in tuple in spite of different order in filter expression)

Comment (ktf) [reply Framework/Core/test/test_ASoAHelpers.cxx:1146]: For this we said that the correct solution would be to have:

```C++
test::X{0} + test::X{1} > 9
```

where (0) and (1) refer to the first and the second table in your combination. This is still not there but I think it should be fairly straighforward, with test::x simply being an instance of `test::X{0}`.

Comment (saganatt) [reply Framework/Core/test/test_ASoAHelpers.cxx:1146]: Yes, with {nr} that makes sense, if we wait with this PR until the filters are upgraded.

**Pull request closed without merging**